File: py/simplechat/simplepub/goset.py
#

# simplepub/goset.py


import hashlib
from collections import deque

from . import node #NODE, LOGTYPE_remote
import time


FID_LEN = 32
GOSET_DMX_STR = "tinySSB-0.1 GOset 1"
DMX_LEN = 7


class Claim:
    typ = b'c'
    lo = bytes(0)  # FID_LEN
    hi = bytes(0)  # FID_LEN
    xo = bytes(0)  # FID_LEN
    sz = 0
    wire = bytes(0)


class Novelty:
    typ = b'n'
    key = bytes(FID_LEN)
    wire = bytes(0)


NOVELTY_LEN = 33 # sizeof(struct novelty_s)
CLAIM_LEN = 98 # sizeof(struct claim_s)
ZAP_LEN = 33 # sizeof(struct zap_s)

GOSET_KEY_LEN   = FID_LEN
GOSET_MAX_KEYS  =    100
GOSET_ROUND_LEN = 10
MAX_PENDING     =     20
ASK_PER_ROUND   =      1
HELP_PER_ROUND  =      3
ZAP_ROUND_LEN   =   4500


class GOset():

    def __init__(self, node, keys=[], verbose=False, log=None) -> None:
        def glob(x): pass
        self.node = node
        self.keys = sorted(keys)
        self.verbose = verbose
        self.log = log if log != None else glob
        self.state = bytes(FID_LEN) if self.keys == [] else \
                     self._xor(0, len(self.keys)-1)
        # self.node.set_want_dmx(self.state)

    goset_dmx = hashlib.sha256(GOSET_DMX_STR.encode()).digest()[:DMX_LEN]

    pending_claims = []
    pending_novelty = deque()
    largest_claim_span = 0

    def _key_to_ndx(self, key) -> int:
        try:
            return self.keys.index(key)
        except:
            return -1

    def incoming_goset_msg(self, dmx, pkt: bytes) -> list:
        lst = []
        if len(pkt) <= DMX_LEN:
            return []
        buf = pkt[DMX_LEN:]
        # print("   =G len:", len(buf), chr(buf[0]))

        if len(buf) == NOVELTY_LEN and buf[0] == ord('n'):
            key = buf[1:NOVELTY_LEN]
            if self.verbose:
                self.log(f"   =G novelty = {key.hex()}")
            self._add_key(key)
            return []
        
        if len(buf) != CLAIM_LEN or buf[0] != ord('c'):
            if self.verbose:
                self.log(f"   =G unknown msg {buf[:1]}")
            return []
        
        cl = self.mkClaim_from_bytes(buf)
        i1,i2 = self._key_to_ndx(cl.lo), self._key_to_ndx(cl.hi)
        if cl.sz == len(self.keys) and self.state == cl.xo:
            syn = "/synced"
        else:
            syn = "/not in sync"
        if self.verbose:
            self.log(f"   =G claim span={cl.sz} lo={i1}:{cl.lo[:10].hex()}.. hi={i2}:{cl.hi[:10].hex()}...  {syn}")

        if cl.sz > self.largest_claim_span:
            self.largest_claim_span = cl.sz
        if i2 - i1 == 1 and cl.sz == 3: # we just lack the key in the middle
            xor = bytearray(cl.xo)
            for i in range(len(xor)):
                xor[i] ^= cl.lo[i]
                xor[i] ^= cl.hi[i]
            if self.verbose:
                self.log(f"   computed the missing middle key {xor.hex()}")
            self._add_key(bytes(xor))
            # FIXME: should we update the state, or do it at ADV time?
        else:
            self._add_key(cl.lo)
            self._add_key(cl.hi)
            self._add_pending_claim(cl)
            # print("   add pending claim")
            # FIXME: should we update the state, or do it at ADV time?
        return []


    def get_adv(self) -> (list, int):
        lst = []
        if len(self.keys) == 0:
            return lst, 15
        cl = self.mkClaim(0, len(self.keys) - 1)
        if cl.xo != self.state:
            if self.verbose:
                self.log(f"   Gadv - new GOset state {cl.xo.hex()} |keys|={len(self.keys)}")
            self.state = cl.xo
            self.node.set_want_dmx(self.state)
        # if len(self.pending_claims) == 0:
        lst.append(self.goset_dmx + cl.wire)
        # return lst, 15

        self.pending_claims.sort(key=lambda x: x.sz)
        max_ask = ASK_PER_ROUND
        max_help = HELP_PER_ROUND

        retain = []
        for c in self.pending_claims:
            if c.sz == 0:
                continue
            lo = self._key_to_ndx(c.lo) # next((i for i, x in enumerate(self.keys) if x == c.lo), -1)
            hi = self._key_to_ndx(c.hi) # next((i for i, x in enumerate(self.keys) if x == c.hi), -1)
            if lo == -1 or hi == -1 or lo > hi:
                continue
            partial = self.mkClaim(lo,hi)
            if partial.xo == c.xo:
                continue
            if partial.sz <= c.sz:
                if max_ask > 0:
                    max_ask -= 1
                    if self.verbose:
                        self.log(f"   Gadv need {lo}..{hi} @{len(self.keys)}")
                    lst.append(self.goset_dmx + partial.wire)
                if partial.sz < c.sz:
                    retain.append(c)
                    continue

            if max_help > 0:
                max_help -= 1
                hi -= 1
                lo += 1
                if hi <= lo:
                    lst.append(self.goset_dmx + self.mkNovelty_from_key(self.keys[lo]).wire)
                elif hi - lo <= 2: # span of 2 or 3
                    if self.verbose:
                        self.log(f"   Gadv have {lo}..{hi} @{len(self.keys)}")
                    lst.append(self.goset_dmx + self.mkClaim(lo,hi).wire)
                else: # split span in two intervals
                    sz = (hi + 1 -lo) // 2
                    if self.verbose:
                        self.log(f"   Gadv have {lo}..{lo+sz-1} @{len(self.keys)}")
                        self.log(f"   Gadv have {lo+sz}..{hi} @{len(self.keys)}")
                    lst.append(self.goset_dmx + self.mkClaim(lo,lo+sz-1).wire)
                    lst.append(self.goset_dmx + self.mkClaim(lo+sz, hi).wire)
                continue
            retain.append(c)
        
        while len(retain) >= MAX_PENDING - 5:
            retain.removeLast()
        self.pending_claims = retain
        return lst, 15


    def _include_key(self, key: bytes) -> None:
        if key == bytes(GOSET_KEY_LEN):
            return False
        if key in self.keys:
            # print("GOset _include_key(): key already exists")
            return False
        if len(self.keys) >= GOSET_MAX_KEYS:
            # print("GOset _include_key(): too many keys")
            return False
        # print("GOset _include_key", key[:8])
        self.keys.append(key)
        return True
            

    def _add_key(self, key: bytes) -> None:
        if key == bytes(GOSET_KEY_LEN) or key in self.keys:
            return
        if len(self.keys) >= GOSET_MAX_KEYS:
            self.log("   too many keys")
            return
        # print(f"   new key {key.hex()}")
        self.keys.append(key)
        self.keys.sort()
        self.node.activate_feed(key)
        '''
        if len(self.keys) >= self.largest_claim_span:
            n = self.mkNovelty_from_key(key)
            if self.novelty_credit > 0:
                self._enqueue(n.wire, self.goset_dmx)
                self.novelty_credit -= 1
            elif len(self.pending_novelty) < MAX_PENDING:
                self.pending_novelty.append(n)
        '''
        self.log(f"   added key {key.hex()}")


    def _add_pending_claim(self, cl: Claim) -> None:
        for c in self.pending_claims:
            if c.sz == cl.sz and c.xo == cl.xo:
                return
        self.pending_claims.append(cl)
    
    '''
    def mkNovelty_from_key(self, key: bytes) -> Novelty:
        n = Novelty()
        n.wire = b'n' + key
        n.key = key
        return n
    '''
    

    def mkClaim_from_bytes(self, pkt: bytes) -> Claim:
        cl = Claim()
        cl.lo = pkt[1:33]
        cl.hi = pkt[33:65]
        cl.xo = pkt[65:97]
        cl.sz = pkt[97]
        cl.wire = pkt
        return cl
    

    def mkClaim(self, lo: int, hi: int) -> Claim:
        cl = Claim()
        cl.lo = self.keys[lo]
        cl.hi = self.keys[hi]
        cl.xo = self._xor(lo, hi)
        cl.sz = hi - lo + 1
        b = bytes([cl.sz])
        cl.wire = cl.typ + cl.lo + cl.hi + cl.xo + b
        return cl


    def _xor(self, lo: int, hi: int) -> bytes:
        if len(self.keys) == 0:
            return bytes(FID_LEN)
        xor = bytearray(self.keys[lo])
        for k in self.keys[lo+1 : hi+1]:
            for i in range(len(xor)):
                xor[i] ^= k[i]
        return xor


    '''
    def adjust_state(self) -> None:
        self.keys.sort()
        if len(self.keys) > 0:
            self.state = self.xor(0,len(self.keys)-1))
        else:
            self.state = bytes(FID_LEN)
        # print("GOset adjust_state() for", len(self.keys), "resulted in", self.state.hex())
        self.node.set_want_dmx(self.state)
    '''

File: android/tinySSB/app/src/main/java/nz/scuttlebutt/tremolavossbol/tssb/GOset.kt
package nz.scuttlebutt.tremolavossbol.tssb

import android.util.Log
import nz.scuttlebutt.tremolavossbol.MainActivity
import nz.scuttlebutt.tremolavossbol.crypto.SodiumAPI.Companion.sha256
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.DMX_LEN
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.FID_LEN
import nz.scuttlebutt.tremolavossbol.utils.Constants.Companion.GOSET_DMX
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.byteArrayCmp
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.fromEncodedUByte
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.toEncodedUByte
import nz.scuttlebutt.tremolavossbol.utils.HelperFunctions.Companion.toHex
import kotlin.experimental.xor

class Claim {
    var typ: Byte = 'c'.toByte()
    var lo = ByteArray(0) // FID_LEN)
    var hi = ByteArray(0) // FID_LEN)
    var xo = ByteArray(0) // FID_LEN)
    var sz = 0
    var wire = ByteArray(0)
}

class Novelty {
    var typ: Byte = 'n'.toByte()
    var key = ByteArray(FID_LEN)
    var wire = ByteArray(0)
}

class GOset(val context: MainActivity) {
    /* packet format:
        n 32B 32B? 32B?  // 33 bytes, in the future up to two additional keys
        c 32B 32B 32B B  // 98 bytes
          lo  hi  xor cnt(4 bytes, big endian)
        z 32B(nonce)     // 33 bytes
    */

    val NOVELTY_LEN = 33 // sizeof(struct novelty_s)
    val CLAIM_LEN = 98   // sizeof(struct claim_s)
    val ZAP_LEN = 33     // sizeof(struct zap_s)

    val GOSET_KEY_LEN   = FID_LEN
    val GOSET_MAX_KEYS  =    100 //   MAX_FEEDS
    val GOSET_ROUND_LEN = 10000L // in millis
    val MAX_PENDING     =     20 // log2(maxSetSize) + density (#neighbors)
    val NOVELTY_PER_ROUND =    1
    val ASK_PER_ROUND   =      1
    val HELP_PER_ROUND  =      2
    val ZAP_ROUND_LEN   =   4500

    val goset_dmx = GOSET_DMX.sha256().sliceArray(0..DMX_LEN-1)

    // unsigned short version; ??
    // const char* fname;
    // int goset_len; // number of set elements
    var state = ByteArray(FID_LEN)
    val keys = ArrayList<ByteArray>(0)
    var pending_claims = ArrayList<Claim>(0)
    val pending_novelty = ArrayDeque<Novelty>(0)
    var largest_claim_span = 0
    var novelty_credit = 1
    // unsigned long next_round; // FIXME: handle, or test, wraparound
    // struct zap_s zap;
    // unsigned long zap_state;
    // unsigned long zap_next;

    fun key2ndx(k: ByteArray): Int {
        return keys.indexOf(k)
    }

    fun rx(pkt: ByteArray, aux: ByteArray?) {
        // Log.d("goset", "rx ${pkt.size}")

        if (pkt.size <= DMX_LEN)
            return
        val buf = pkt.sliceArray(DMX_LEN .. pkt.lastIndex)

        if (buf.size == NOVELTY_LEN && buf[0] == 'n'.toByte()) {
            _add_key(buf.sliceArray(1..NOVELTY_LEN-1))
            return
        }
        /*
        if (pkt[0] == 'z' && len == ZAP_LEN) {
            Serial.println("ZAP message received");
            unsigned long now = millis();
            if (gp->zap_state == 0) {
                Serial.println("ZAP phase I starts");
                memcpy(&gp->zap, pkt, ZAP_LEN);
                gp->zap_state = now + ZAP_ROUND_LEN;
                gp->zap_next = now;
            }
            return;
        }
        */
        if (buf.size != CLAIM_LEN || buf[0] != 'c'.toByte())
            return;
        val cl = mkClaim(buf)

        if (cl.sz > largest_claim_span)
            largest_claim_span = cl.sz
        if (context.tinyRepo.isLoaded())
            context.wai.eval("refresh_goset_progressbar(${keys.size}, ${largest_claim_span})") // notify frontend
        // Log.d("rx", "state = ${state.toHex()}, cl.sz=${cl.sz}, k.sz=${keys.size}")
        if (cl.sz == keys.size && byteArrayCmp(state, cl.xo) == 0) {
            Log.d("goset", "seems we are synced (with at least someone), |GOset|=${keys.size}")
        } else {
            val i1 = keys.indexOf(cl.lo)
            val i2 = keys.indexOf(cl.hi)
            if (i2-i1 == 1 && cl.sz == 3) {
                for (i in cl.xo.indices) {
                    cl.xo[i] = cl.xo[i] xor cl.lo[i]
                    cl.xo[i] = cl.xo[i] xor cl.hi[i]
                }
                Log.d("goset", "computed the mussung middle key ${cl.xo.toHex()}")
                _add_key(cl.xo)
            } else {
                _add_key(cl.lo)
                _add_key(cl.hi)
                _add_pending_claim(cl)
            }
        }
        // Log.d("goset", "goset size=${keys.size}, pending_claims.size=${pending_claims.size}")
    }

    fun loop() {
        while (true) {
            beacon()
            Thread.sleep(GOSET_ROUND_LEN)
        }
    }

    fun beacon() { // called in regular intervals
        /*
        unsigned long now = millis();
        if (gp->zap_state != 0) {
            if (now > gp->zap_state + ZAP_ROUND_LEN) { // two rounds after zap started
            Serial.println("ZAP phase II ended, resetting now");
            repo_reset();
        }
            if (now < gp->zap_state && now > gp->zap_next) { // phase I
            Serial.println("sending zap message");
            io_send(_mkZap(gp), DMX_LEN + ZAP_LEN, NULL);
            gp->zap_next = now + 1000;
        }
        }
        */
        // Log.d("GOset", "beacon")
        if (keys.size == 0 || !context.tinyRepo.isLoaded()) return
        while (novelty_credit-- > 0 && pending_novelty.size > 0)
            context.tinyIO.enqueue(pending_novelty.removeFirst().wire, goset_dmx, null)
        novelty_credit = NOVELTY_PER_ROUND

        val cl = mkClaim(0, keys.size-1)
        if (byteArrayCmp(cl.xo, state) != 0) { // GOset changed
            Log.d("goset", "state change to ${cl.xo.toHex()}, |keys|=${keys.size}")
            state = cl.xo
            context.tinyDemux.set_want_dmx(state)
        }
        context.tinyIO.enqueue(cl.wire, goset_dmx, null);

        // sort pending entries, smallest first
        pending_claims.sortBy({c -> c.sz})
        var max_ask = ASK_PER_ROUND
        var max_help = HELP_PER_ROUND

        val retain = ArrayList<Claim>()
        for (c in pending_claims) {
            Log.d("goset", "pending claim loop")
            if (c.sz == 0) continue // ignore bogous claims
            var lo = keys.indexOfFirst({k -> byteArrayCmp(k,c.lo) == 0})
            var hi = keys.indexOfFirst({k -> byteArrayCmp(k,c.hi) == 0})
            if (lo == -1 || hi == -1 || lo > hi) continue // ignore bogous claims
            val partial = mkClaim(lo, hi)
            if (byteArrayCmp(partial.xo, c.xo) == 0) continue // match, hence no reason to ask others
            if (partial.sz <= c.sz) { // ask for help, starting with the smallest entry, for a limited number of questions
                if (max_ask-- > 0) {
                    // Serial.print("asking for help " + String(partial->cnt));
                    // Serial.print(String(" ") + to_hex(partial->lo,4) + String(".."));
                    // Serial.println(String(" ") + to_hex(partial->hi,4) + String(".."));
                    context.tinyIO.enqueue(partial.wire, goset_dmx, null);
                }
                if (partial.sz < c.sz) {
                    retain.add(c)
                    continue;
                }
            }
            if (max_help-- > 0) { // we have larger claim span, offer help (but limit # of claims)
                hi--; lo++
                // Log.d("goset","offer help span= " + String(partial->cnt - 2));
                // Serial.print(String(" ") + to_hex(gp->goset_keys+lo*GOSET_KEY_LEN,4) + String(".."));
                // Serial.println(String(" ") + to_hex(gp->goset_keys+hi*GOSET_KEY_LEN,4) + String(".."));
                Log.d("goset", "max_help ${lo}, ${hi}")
                if (hi <= lo)
                    context.tinyIO.enqueue(mkNovelty_from_key(keys[lo]).wire, goset_dmx, null)
                else if (hi - lo <= 2) { // span of 2 or 3
                    context.tinyIO.enqueue(mkClaim(lo, hi).wire, goset_dmx, null)
                }
                else { // split span in two intervals
                    val sz = (hi+1 - lo) / 2
                    context.tinyIO.enqueue(mkClaim(lo, lo+sz-1).wire, goset_dmx, null)
                    context.tinyIO.enqueue(mkClaim(lo+sz, hi).wire, goset_dmx, null)
                }
                continue; // only help once per claim, do not retain
            }
            retain.add(c) // we could not treat the claim in this round, hence retain
        }
        // make room for new claims
        while (retain.size >= MAX_PENDING-5)
            retain.removeLast()
        pending_claims = retain
        // Serial.println(String(gp->pending_c_cnt,DEC) + " pending claims, GOset size is " + String(gp->goset_len, DEC));
        // for (int i = 0; i < gp->pending_c_cnt; i++)
        // Serial.println("  pend claim span=" + String(gp->pending_claims[i].cnt, DEC)
        // + " " + to_hex(gp->pending_claims[i].xo,32));
    }

    fun _include_key(key: ByteArray): Boolean {
        val zero = ByteArray(GOSET_KEY_LEN)
        if (key.contentEquals(zero))
            return false
        for (k in keys)
            if (k.contentEquals(key)) {
                // Log.d("goset", "key ${key.toHex()} already exists")
                return false
            }
        if (keys.size >= GOSET_MAX_KEYS) {
            Log.d("goset", "too many keys: ${keys.size}")
            return false
        }
        Log.d("goset","_include_key ${key.toHex()}, keys.size was ${keys.size}")
        keys.add(key)
        return true
    }

    fun _add_key(key: ByteArray, trusted: Boolean = false, alias: String =""): String {
        if (!_include_key(key)) // adds key if necessary
            return ""
        //Set trusted
        context.tinyRepo.add_replica(key, trusted, alias)

        keys.sortWith({a:ByteArray,b:ByteArray -> byteArrayCmp(a,b)})
        if (keys.size >= largest_claim_span) { // only rebroadcast if we are up to date
            val n = mkNovelty_from_key(key)
            if (novelty_credit-- > 0)
                context.tinyIO.enqueue(n.wire, goset_dmx, null)
            else if (pending_novelty.size < MAX_PENDING)
                pending_novelty.add(n)
        }
        context.ble?.refreshShortNameForKey(key) // refresh shortname in devices overview
        //Set
        Log.d("goset", "added key ${key.toHex()}, |keys|=${keys.size}")
        if (!trusted) {
            //add entry to json file with the hex
            context.saveContactToJSONFile(key.toHex(), 24)
        }
        return key.toHex()
    }

    fun remove_key(key: ByteArray) {
        for (k in keys) {
            if (k.contentEquals(key)) {
                keys.remove(k)
                Log.d("goset", "removed key ${key.toHex()}, |keys|=${keys.size}")
                break
            }
        }
    }

    fun _add_pending_claim(cl: Claim) {
        for (c in pending_claims)
            if (c.sz == cl.sz && c.xo.contentEquals(cl.xo))
                return
        pending_claims.add(cl)
    }

    fun adjust_state() {
        keys.sortWith({a:ByteArray,b:ByteArray -> byteArrayCmp(a,b)})
        if (keys.size > 0) {
            val cl = mkClaim(0, keys.size-1)
            state = cl.xo
        } else
            state = ByteArray(FID_LEN)
        Log.d("goset", "adjust state for ${keys.size} keys, resulted in ${state.toHex()}")
        context.tinyDemux.set_want_dmx(state)
    }

    fun mkClaim(pkt: ByteArray): Claim {
        val cl = Claim()
        cl.lo = pkt.sliceArray(1..32)
        cl.hi = pkt.sliceArray(33..64)
        cl.xo = pkt.sliceArray(65..96)
        cl.sz = pkt[97].fromEncodedUByte()
        cl.wire = pkt
        return cl
    }

    fun mkClaim(lo: Int, hi: Int): Claim {
        val cl = Claim()
        cl.lo = keys[lo]
        cl.hi = keys[hi]
        cl.xo = _xor(lo, hi)
        cl.sz = hi - lo + 1
        val b = cl.sz.toEncodedUByte()
        cl.wire = byteArrayOf(cl.typ) + cl.lo + cl.hi + cl.xo + byteArrayOf(b)
        return cl
    }

    fun mkNovelty(pkt: ByteArray): Novelty {
        val n = Novelty()
        n.key = pkt.sliceArray(1..33)
        n.wire = pkt
        return n
    }

    fun mkNovelty_from_key(key: ByteArray): Novelty {
        val n = Novelty()
        n.wire = byteArrayOf('n'.toByte()) + key
        n.key = key
        return n
    }

    fun _xor(lo: Int, hi: Int): ByteArray {
        val xor = keys[lo].clone()
        for (k in keys.subList(lo+1,hi+1)) // argh, in Kotlin the toIndex is EXCLUSIVE, unlike in a range
            for (i in 0..FID_LEN-1)
                xor[i] = xor[i] xor k[i]
        // Log.d("xor", "lo=${keys[lo].toHex()}, cnt=${keys.size}")
        // Log.d("xor", "hi=${keys[hi].toHex()}")
        // Log.d("xor", "xo=${xor.toHex()}")
        return xor
    }

}